console/src/state/resources.rs

```diff
@@ -0,0 +1,322 @@
+                    let dropped_for = now.duration_since(d).unwrap();
+                    retain_for > dropped_for
+                })
+                .unwrap_or(true)
+        })
+    }
+}
+
+impl Resource {
+    pub(crate) fn id(&self) -> u64 {
+        self.id
+    }
+
+    pub(crate) fn target(&self) -> &str {
+        &self.target
+    }
+
+    pub(crate) fn concrete_type(&self) -> &str {
+        &self.concrete_type
+    }
+
+    pub(crate) fn kind(&self) -> &str {
+        match &self.kind {
+            Kind::Timer => "Timer",
+            Kind::Other(other) => other,
+        }
+    }
+
+    pub(crate) fn formatted_attributes(&self) -> &[Vec<Span<'static>>] {
+        &self.stats.formatted_attributes
+    }
+
+    pub(crate) fn total(&self, since: SystemTime) -> Duration {
+        self.stats
+            .total
+            .unwrap_or_else(|| since.duration_since(self.stats.created_at).unwrap())
+    }
+
+    pub(crate) fn dropped(&self) -> bool {
+        self.stats.total.is_some()
+    }
+
+    pub(crate) fn location(&self) -> String {
+        if let Some(location) = self.location.as_ref() {
+            return format!("{} ", location);
+        }
+        "unknown location".to_string()
+    }
+}
+
+impl ResourceStats {
+    fn from_proto(pb: proto::resources::Stats, meta: &Metadata, styles: &view::Styles) -> Self {
+        let mut pb = pb;
+        let mut attributes = pb
+            .attributes
+            .drain(..)
+            .filter_map(|pb| {
+                let field = pb.field?;
+                let field = Field::from_proto(field, meta)?;
+                Some(Attribute {
+                    field,
+                    unit: pb.unit,
+                })
+            })
+            .collect::<Vec<_>>();
+
+        let formatted_attributes = Attribute::make_formatted(styles, &mut attributes);
+        let created_at = pb
+            .created_at
+            .expect("task span was never created")
+            .try_into()
+            .unwrap();
+        let dropped_at: Option<SystemTime> = pb.dropped_at.map(|v| v.try_into().unwrap());
+        let total = dropped_at.map(|d| d.duration_since(created_at).unwrap());
+
+        Self {
+            created_at,
+            dropped_at,
+            total,
+            attributes,
+            formatted_attributes,
+        }
+    }
+}
+
+impl From<proto::resources::resource::Kind> for Kind {
+    fn from(pb: proto::resources::resource::Kind) -> Self {
+        use proto::resources::resource::kind::Kind::Known as PbKnown;
+        use proto::resources::resource::kind::Kind::Other as PBOther;
+        use proto::resources::resource::kind::Known::Timer as PbTimer;
+
+        match pb.kind.expect("a resource should have a kind field") {
+            PbKnown(known) if known == (PbTimer as i32) => Kind::Timer,
+            PbKnown(known) => panic!("failed to parseknown kind from {}", known),
+            PBOther(other) => Kind::Other(other),
+        }
+    }
+}
```

Comment:
nit, take it or leave it: should this maybe be a `TryFrom` implementation, rather than just panicking? it might be better to just log and ignore resources with messed up kinds?